src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -513,171 +517,224 @@ private static String GetFullSearchString(String fullPath, String searchPattern)
-                return di;
-            }
-            else
-            {
-                Debug.Assert(isFile);
-                IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(name, result.FindData, asDirectory: false);
-                FileInfo fi = new FileInfo(name, fileSystemObject);
-                return fi;
+                IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(fullPath, findData, asDirectory: true);
+                result = new DirectoryInfo(fullPath, fileSystemObject);
+                return true;
             }
+
+            result = null;
+            return false;
         }
     }
 
-    // Holds search data that is passed around 
-    // in the heap based stack recursion
-    internal sealed class SearchData
+    internal class FileSystemInfoResultHandler : SearchResultHandler<FileSystemInfo>
     {
-        public SearchData(String fullPath, String userPath, SearchOption searchOption)
+        [System.Security.SecurityCritical]
+        internal override bool IsResultIncluded(string fullPath, string userPath, ref Interop.mincore.WIN32_FIND_DATA findData, out FileSystemInfo result)
         {
-            Contract.Requires(fullPath != null && fullPath.Length > 0);
-            Contract.Requires(userPath != null && userPath.Length > 0);
-            Contract.Requires(searchOption == SearchOption.AllDirectories || searchOption == SearchOption.TopDirectoryOnly);
+            bool isFile = Win32FileSystemEnumerableHelpers.IsFile(ref findData);
+            bool isDir = Win32FileSystemEnumerableHelpers.IsDir(ref findData);
+            Debug.Assert(!(isFile && isDir), findData.cFileName + ": current item can't be both file and dir!");
 
-            this.fullPath = fullPath;
-            this.userPath = userPath;
-            this.searchOption = searchOption;
-        }
+            if (isFile)
+            {
+                IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(fullPath, findData, asDirectory: false);
+                result = new FileInfo(fullPath, fileSystemObject);
+                return true;
+            }
+            else if (isDir)
+            {
+                IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(fullPath, findData, asDirectory: true);
+                result = new DirectoryInfo(fullPath, fileSystemObject);
+                return true;
+            }
 
-        public readonly string fullPath;     // Fully qualified search path excluding the search criteria in the end (ex, c:\temp\bar\foo)
-        public readonly string userPath;     // User specified path (ex, bar\foo)
-        public readonly SearchOption searchOption;
+            result = null;
+            return false;
+        }
     }
 
-    internal sealed class SearchResult
+    // Holds search data that is passed around
+    // in the heap based stack recursion
+    internal struct SearchData
```

Comment:
With this now a struct and with only the two fields, it's now the equivalent of the ```PathPair``` type in the Unix version of this: https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs#L527-L537 .  It'd be nice to separate this out and unify them at some point.